README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->
<!-- badges: start -->

[![CRAN
status](https://www.r-pkg.org/badges/version/tabularaster)](https://CRAN.R-project.org/package=tabularaster)
[![Lifecycle:
stable](https://img.shields.io/badge/lifecycle-stable-brightgreen.svg)](https://lifecycle.r-lib.org/articles/stages.html#stable)
[![R-CMD-check](https://github.com/hypertidy/tabularaster/actions/workflows/R-CMD-check.yaml/badge.svg)](https://github.com/hypertidy/tabularaster/actions/workflows/R-CMD-check.yaml)
<!-- badges: end -->

# tabularaster

The `raster` package is extremely powerful for spatial data. It provides
very efficient data extraction and summary tools via consistent
cell-index and comprehensive set of functions for working with grids,
cells and their values.

Tabularaster provides some more helpers for working with cells and tries
to fill some of the (very few!) gaps in raster functionality. When
raster returns cell values of hierarchical objects it returns a
hierarchical (list) of cells to match the input query, while
`tabularaster::cellnumbers` instead returns a data frame of identifiers
and cell numbers.

Tabularaster provides these functions.

-   `as_tibble` - convert to data frame with options for value column
    and cell, dimension and date indexing
-   `cellnumbers` - extract of cell index numbers as a simple data frame
    with “object ID” and “cell index”
-   `index_extent` - create an index extent, essentially
    `extent(0, ncol(raster), 0, nrow(raster))`

All functions that work with `sp Spatial` also work with
`sf simple features`.

There is some overlap with `quadmesh` and `spex` while I figure out
where things belong.

# Installation

Install from CRAN,

``` r
install.packages("tabularaster")
```

or get the development version from Github.

``` r
devtools::install_github("hypertidy/tabularaster")
```

# Usage

Basic usage is to extract the cell numbers from an object, where object
is a a matrix of points, a `Spatial` object or a `simple features sf`
object.

``` r
cells <- cellnumbers(raster, object)
```

The value in this approach is not for getting cell numbers per se, but
for using those downstream. The cell number is an index into the raster
that means the geometric hard work is done, so we can apply the index
for subsequent extractions, grouping aggregations, or for determining
the coordinates or other structure summaries of where the cell belongs.

E.g.

``` r

## summarize by object grouping
cells %>% mutate(value= extract(raster, cell_)) %>% group_by(object_) %>% summarize(mean(value))

## summarize by cell grouping
cells %>% mutate(value= extract(raster, cell_)) %>% group_by(cell_) %>% summarize(mean(value))
```

The utility of this is very much dependent on individual workflow, so
this in its own right is not very exciting: Tabularaster simply provides
an easier way to create news tools when format vagaries aren’t your day
job.

See the vignettes for more.

------------------------------------------------------------------------

Please note that the ‘silicate’ project is released with a [Contributor
Code of
Conduct](https://github.com/hypertidy/tabularaster/blob/master/CODE_OF_CONDUCT.md).
By contributing to this project, you agree to abide by its terms.


NEWS.md
# tabularaster 0.7.2

* Fixed namespace doc thanks to CRAN.

# tabularaster 0.7.1

* Fix thanks to CRAN, now handling new Date and POSIXt behaviour with integer input #31. 

* More robust to broken raster date-time values, #30. 

# tabularaster 0.7.0

* Fix randomizing order for sf `object_`, thanks to @CaptureOSM https://github.com/hypertidy/tabularaster/issues/29. 

* Align to new spatstat 2.0.0.

# tabularaster 0.6.6

* Release to restore package on CRAN, archived when silicate was archived (for neglecting
 conditional use of Suggests). 
 
# tabularaster 0.6.5

* Fixed unnecessary test for warning contributed by R. Hijmans. 

# tabularaster 0.6.0

* Removed unnecessary issues with crs metadata strings. 

* Now using silicate for the line-to-segments ops, removed spbabel imports. 

* Remove old `bufext()` and outdated spex import. 

* Un-deprecate and re-export `decimate()`. 

* Drop `extentFromCells()` export. 

# tabularaster 0.5.0

* Function `cellnumbers` now provides only basic cell lookup, the weights and other
 options available in raster functions are ignored. 
 
* Function `extentFromCells` is now imported from raster and re-exported. 

* Now applying `fasterize` for sf polygon objects, and `spatstat` for sf line objects for 
 very significant speed ups. 

* Now import `spatstat` for faster line rasterization. 

# tabularaster 0.4.0

* add `xy` argument to `as_tibble`

* fix missing vignette knitr reference

* new oisst data set

# tabularaster 0.3.0

* bufext made defunct

* added index extent functions `index_extent` and `cellsFromExtent`

* removed decimate function

* fixed unnecessary lapply call for the points case, so that is now not slow

* extra handling for as_tibble, value is now optional so that data-less rasters can be used to 
 build cell and dimension indices
 
* new vignette stub to list all relevant raster functions

* object_ is now integer, 1-based to match the rows in query

* removed `boundary` (see `romsboundary` in `angstroms`)

* consolidated on the central tools, removed experiments

* migrated the buffer extent logic to `spex`, deprecated `bufext`

* added `as_tibble` for rasters

* added boundary function

# tidyraster 0.2.0

* new data sets rastercano and polycano

* new functions decimate, cellnumbers and bufext

* Added a `NEWS.md` file to track changes to the package.





R/tibble.R
#' Convert a Raster to a data frame. 
#'
#' Generate a data frame version of any raster object. Use the arguments
#' 'cell', 'dim', 'split_date' and 'value' to control the columns that
#' are included in the output.
#' 
#' If the raster has only one layer, the slice index is not added. Use 'dim = FALSE' to not include
#' the slice index value. 
#'
#' @param x a RasterLayer, RasterStack or RasterBrick
#' @param cell logical to include explicit cell number
#' @param dim logical to include slice index 
#' @param split_date logical to split date into components
#' @param value logical to return the values as a column or not
#' @param xy logical to include the x and y centre coordinate of each cell
#' @param ... unused
#'
#' @return a data frame (tibble) with columns: 
#' 
#' * `cellvalue` the actual value of the raster cell
#' * `cellindex` the index of the cell (numbered from 1 to `ncell()` in the raster way). 
#' 
#' Columns `cellindex` or `cellvalue` may be omitted if either or both of `cell` and/or `value` are `FALSE`, respectively
#' 
#' Other columns might be included depending on the properties of the raster and
#' the arguments to the function: 
#' 
#' * `year`,`month`,`day` if `split_date` is `TRUE`
#' * `x`,`y` if `xy` is `TRUE`
#' * `dimindex` if the input has more than 1 layer and `dim` is `TRUE`. 
#' 
#' 
#' @export
#'
#' @examples
#' ## basic data frame version of a basic raster
#' as_tibble(raster::raster(volcano))
#' 
#' ## data frame with time column since raster has that set
#' r <- raster::raster(volcano)
#' br <- raster::brick(r, r)
#' as_tibble(raster::setZ(br, Sys.Date() + 1:2), cell = TRUE)
#' @importFrom tibble as_tibble tibble
#' @export as_tibble
#' @importFrom dplyr bind_cols mutate
#' @importFrom raster getZ nlayers values ncell
#' @name as_tibble
as_tibble.BasicRaster <- function(x, cell = TRUE, dim = nlayers(x) > 1L, value = TRUE, split_date = FALSE, xy = FALSE,  ...) {
  dimindex <- raster::getZ(x)

  if (!(inherits(dimindex, "Date") || inherits(dimindex, "POSIXt"))) {
    ## we can't do this if Z isn't this class, because R 4.3 allows as.Date(1) for example #31
    split_date <- FALSE
  }
  if (is.null(dimindex)) {
    dimindex <- seq(raster::nlayers(x))
    if (split_date) {
      e1 <- try(as.Date(dimindex), silent = TRUE)
      e2 <- try(as.POSIXct(dimindex, tz = "GMT"), silent = TRUE)
      if ((inherits(e1, "try-error") & inherits(e2, "try-error")) | any(is.na(range(e1)))) {
        message("cannot 'split_date', convert 'getZ(x)' not convertible to a Date or POSIXct")
        split_date <- FALSE
      }
    }
  } else {
    if (length(dimindex) != raster::nlayers(x)) {
      ## warning message in case of #30 
      warning(sprintf("raster object is invalid, length of getZ(x): '%i' should match nlayers(x): '%i'", length(raster::getZ(x)), raster::nlayers(x)))
    }
  }
  
  cellvalue <- cellindex <-  NULL
  if (value) cellvalue <- as.vector(raster::values(x))
  cellindex <-  rep(seq(raster::ncell(x)), raster::nlayers(x))
  d <- dplyr::bind_cols(cellvalue = cellvalue, cellindex = cellindex)
  if (dim) {
    dimindex <- rep(dimindex, each = ncell(x))
    if (split_date && inherits(dimindex, "Date")) {
      d <- dplyr::mutate(d, year = as.integer(format(dimindex, "%Y")), 
                                month = as.integer(format(dimindex, "%m")), 
                                day = as.integer(format(dimindex, "%d")))
      
    } else {
      d[["dimindex"]] <- dimindex
    }
  }
  if (xy) {
    xy <- raster::xyFromCell(x, seq_len(raster::ncell(x)))
    colnames(xy) <- c("x", "y")
    if (nlayers(x) > 1) xy <- xy[rep(seq_len(ncell(x)), nlayers(x)), ]
    d <- dplyr::bind_cols(d, tibble::as_tibble(xy))
  }
  if (!cell) d[["cellindex"]] <- NULL
  d
}

